### PR TITLE
MySQL proxy handshake fix

### PR DESCRIPTION
**Pull request by r0mant**
Update MySQL proxy to recover from panic in case the client sends invalid handshake message. Refs https://github.com/gravitational/cloud/issues/1111.

### CODE REVIEW

**Review by timothyb89 — APPROVED** [review]
I noticed @knisbet linked to an upstream fix for this particular panic in the linked cloud issue. For my own curiosity, what's the thought process behind going the panic recovery route? Is it meant to be a short term fix?

**Comment by r0mant** [comment]
> I noticed @knisbet linked to an upstream fix for this particular panic in the linked cloud issue. For my own curiosity, what's the thought process behind going the panic recovery route? Is it meant to be a short term fix?

@timothyb89 The fix upstream is actually also a panic recovery: https://github.com/go-mysql-org/go-mysql/pull/557. The reason I didn't want to just take that in is when writing the unit tests I discovered that panic can also happen in other parts of the handshake which the upstream fix doesn't cover so I implemented it in teleport instead. I will re-vendor mysql library too just to bring it up-to-date when I have time but that would be a separate undertaking.

File: lib/srv/db/mysql/proxy.go [hunk 1]
```diff
@@ -61,6 +61,10 @@ func (p *Proxy) HandleConnection(ctx context.Context, clientConn net.Conn) (err
 	// If any error happens, make sure to send it back to the client so it
 	// has a chance to close the connection from its side.
 	defer func() {
+		if r := recover(); r != nil {
+			p.Log.Warnf("Recovered in MySQL proxy while handling connection from %v: %v.", clientConn.RemoteAddr(), r)
+			err = trace.BadParameter("failed to handle MySQL client connection")
```
Comment (knisbet) [inline lib/srv/db/mysql/proxy.go:66]: Nit: can we get additional fields about the client such as source IP etc? So if someone is targetting us we can use the logs to identify the source?

Comment (r0mant) [reply lib/srv/db/mysql/proxy.go:66]: @knisbet The log entry includes client remote addr but that's probably not very useful if there's a load balancer. We also do log proxy line if lb sends it: https://github.com/gravitational/teleport/blob/12063b2cfe6c190823a4d4b9c8733ed3a0630821/lib/srv/db/mysql/proxy.go#L175

Would that work? I could also see if we can propagate the source client info from proxy line to this handler if that's better.

**Review by smallinsky — APPROVED** [review]

File: lib/srv/db/mysql/proxy.go (hunk 1 above)
Comment (knisbet) [reply lib/srv/db/mysql/proxy.go:66]: Yea, that's a good point. I think the proxy-line would work, as we send proxy-protocol from our edge to the proxies.

**Pull request merged**